### PR TITLE
metrics: Submit metrics when exiting.  Refactor `MetricsAgent::run()`.

### PR DESCRIPTION
**Pull request by illia-bobyr**
There are a few minor issues this change addresses:

1. When we send points to the `MetricsWriter` we are calling `Instant::now()` twice, using the first result in the metrics stats, and using the seconds value for `last_write_time`.  Yet, on the next upload, we would use `last_write_time` as a reference point.

   We upload metrics using a network call, so it is far from instantaneous.  This creates a minor discrepancy in our time reporting.

   Good news is that we do not really need to call `Instant::now()` twice at all, as we can use the same value for both stats and `last_write_time`.

2. We did not report metrics stats if we did not have any points accumulated.  It seems better to always report metric stats, including when no points have been accumulated.  In practice, this does not happen for the validator, as validators always report something during a 10-second accumulation interval.

3. We did not upload any points when the metrics thread was existing. This may cause a short number of metrics not to be reported.

4. `collect_points()` was always converting both `points` and `counters` into a vector of `DataPoint`, even if the final length was over the specified `max_points`.  In the `mainnet-beta` we have values of up to 5m points lost, so it could be a small optimization if we drop them sooner.

### CODE REVIEW

File: metrics/src/metrics.rs [hunk 1]
```diff
@@ -630,7 +663,7 @@ mod test {
         let writer = Arc::new(MockMetricsWriter::new());
         let agent = MetricsAgent::new(writer.clone(), Duration::from_secs(1), 100);
 
-        for i in 0..102 {
+        for i in 0..110 {
```
Comment (bw-solana) [inline metrics/src/metrics.rs]: Is the point here just that 110 is > 100? Might be nice to use a constant for the rate and then make this some offset of the max to make it more obvious what we're doing. Could also use the max below in the assert

**Review by bw-solana — DISMISSED** [review]
LGTM. Minor comment about the test

File: metrics/src/metrics.rs (hunk 1 above)
Comment (illia-bobyr) [reply metrics/src/metrics.rs]: Yes, it is just larger and the difference is more than just 1 or 2, as now both `submit()` and `flush()` upload one "extra" entry that holds the metrics stats.

Good point on adding a constant.  Added.
I think it is more readable this way.

**Review by bw-solana — DISMISSED** [review]
LGTM

**Review by bw-solana — APPROVED** [review]

**Pull request merged**